6LoWPAN Fragmentation

ptcrews opened the pull request:
Hi all,

This is an initial attempt at the 6LoWPAN fragmentation layer. Currently, this layer sits above the MAC layer, and is responsible for keeping track of the transmit/receive state for sending IPv6 packets. This layer will fragment packets if necessary, and also performs 6LoWPAN compression/decompression when necessary. A more formal description of the code is included in the `capsules/src/net/lowpan_fragment.rs` file. Note that there are still several problems with this current code:
- Needs to allow for optional 6LoWPAN compression (currently, all packets are compressed/decompressed using 6LoWPAN)
- During testing, sometimes the receive path stops printing debug messages to the console
- The ReceiveClient trait should receive an immutable reference, rather than a mutable one
- This layer should probably expose a ConfigClient interface to configure radio/MAC layer state
- This code has *not* been tested using the new Imix platform - it has only been tested on the Imixv1 platform

Note that this code has passed basic fragmentation/reassembly testing on the Imixv1 platform, but still requires substantial testing on the Imix(v2) platform. Also note that the main.rs file has been modified to include initialization code - I'm not sure what form this file will take in the end, but there could probably be substantial clean-up that could make `main.rs` more readable.

Please let us know if there are any changes we should make!

Thanks!

alevy commented on boards/imix/chip_layout.ld:
```diff
@@ -1,8 +1,8 @@
-ROM_LENGTH  = 0x00020000;
-PROG_ORIGIN = 0x00030000;
-PROG_LENGTH = 0x00040000;
+ROM_LENGTH  = 0x00025000;
```
Sigh... this is kind of a bummer, but for sanity, I compiled open thread for the CC2650 and it looks like it's on the order of 86-105kB for _just_ open thread (obviously with no sensor drivers or USB or anything), and this is still under half of flash for what's a basically very featureful kernel.

alevy commented on boards/imix/src/lowpan_frag_dummy.rs:
```diff
@@ -0,0 +1,507 @@
+//! A dummy sixlowpan/IP sender
```
What's the difference between this file and `sixlowpan_dummy.rs`?

ptcrews replied:
They are fairly similar, but `sixlowpan_dummy.rs` only tests compression/decompression, whereas `lowpan_frag_dummy.rs` tests fragmentation and reassembly, while the lower layer (the implemented fragmentation layer) does the compression invisibly to the test suite. I think it is useful to have both types of tests as it would be easier to pinpoint whether a bug is in the compression/decompression routines versus the fragmentation layer, but they could probably be combined into a single file.

alevy commented on boards/imix/src/main.rs:
```diff
@@ -8,6 +8,9 @@ extern crate compiler_builtins;
 extern crate sam4l;
 
+use capsules::mac::Mac;
+#[allow(unused_imports)]
```
I don't think `unused_imports` is necessary here

alevy commented on boards/imix/src/main.rs:
```diff
@@ -395,19 +406,76 @@ pub unsafe fn reset_handler() {
+    tx_state.set_transmit_client(lowpan_frag_dummy);
+    frag_state_alarm.set_client(frag_state);
+    frag_dummy_alarm.set_client(lowpan_frag_dummy);
+    frag_state.schedule_next_timer();
```
OK, so either we should have not include this capsule in `imix` by default, or it should be called something other than `*_dummy`.

alevy commented on boards/imix/src/main.rs:
```diff
@@ -441,11 +509,9 @@ pub unsafe fn reset_handler() {
-    rf233.set_address(0x1008);
-    //    rf233.config_commit();
-
     rf233.start();
```
seems like `reset()` and `start()` ought to be in the Mac layer (or whichever component is the immediate client of `rf233`).

ptcrews replied:
So we're not entirely sure why, but the `reset()` and `start()` methods need to be below the creation of the chip otherwise the radio will not work, so I don't think they can be cleanly added to the current Mac initialization code. We could modify the Mac layer to expose a wrapper for `reset()` and `start()`, or we can try and figure out how to incorporate it more cleanly with the lower layers.

In either case, I'm not sure if we should implement it in this pull request, or submit another (small) pull request that modifies the Mac layer.

alevy commented on boards/imix/src/sixlowpan_dummy.rs:
```diff
@@ -0,0 +1,454 @@
+//! A dummy sixlowpan/IP sender
```
A bit more descriptive comment

alevy commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -0,0 +1,863 @@
+        if ip6_packet_option.is_none() {
+            return Err((ReturnCode::ENOMEM, frag_buf));
+        }
+        let ip6_packet = ip6_packet_option.unwrap();
```
Shouldn't use `unwrap` if avoidable. In this case, i would restructure it like this:

```rust
match ip6_packet_option {
    None => return Err((ReturnCode::ENOMEM, frag_buf)),
    Some(ip6_packet) => {
         ...rest of method...
    }
}
```

alevy commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -0,0 +1,863 @@
+        // a callback for an invalid frame reception
+        let data_offset = data_offset;
+        // TODO: Handle the case where the addresses are None/elided
+        let src_mac_addr = header.src_addr.unwrap();
```
There are _a lot_ of `unwrap`s all over this PR. They should basically all be removed in favor of `map`, `match` or `if let Some(my_var) = my_option`. If there are places where it really really really doesn't make sense to have a case where the variable is `None`, either reconsider the type of the variable/field, or at the very least, replace it with `expect` and meaningful error message describing why it's an impossible result.

ptcrews replied:
There's a couple places where I'm not sure what the desired behavior should be. For example, in when sending a new packet and the `tx_buf` field of the `TxState` struct is `None`, then we currently panic. We don't *need* to panic here, but this case should not happen (and therefore represents a logic bug), and even if we return a failure state instead of panicking, we can never send using the same `TxState` struct (as the buffer is now permanently lost). There are 2 more analogous cases I can identify in the code. Should the code "fail fast" and panic (by calling `expect`) in these cases, or is it preferable to keep running in an inconsistent state?

phil-levis commented:
I’m of the opinion that form logic bugs that will prevent future operation we should fail fast. In the best case, Tock should log the failure to flash and reboot.

Phil

> On Aug 30, 2017, at 11:28 AM, ptcrews <notifications@github.com> wrote:
> 
> @ptcrews commented on this pull request.
> 
> In capsules/src/net/lowpan_fragment.rs:
> 
> > +                    self.end_packet_transmit(acked, retcode);
> +                    self.tx_buf.replace(ret_buf);
> +                });
> +            }
> +        });
> +    }
> +}
> +
> +// This function is called after receiving a frame
> +impl<'a, A: time::Alarm> RxClient for FragState<'a, A> {
> +    fn receive<'b>(&self, buf: &'b [u8], header: Header<'b>, data_offset: usize, data_len: usize) {
> +        // We return if retcode is not valid, as it does not make sense to issue
> +        // a callback for an invalid frame reception
> +        let data_offset = data_offset;
> +        // TODO: Handle the case where the addresses are None/elided
> +        let src_mac_addr = header.src_addr.unwrap();
> 
> There's a couple places where I'm not sure what the desired behavior should be. For example, in when sending a new packet and the tx_buf field of the TxState struct is None, then we currently panic. We don't need to panic here, but this case should not happen (and therefore represents a logic bug), and even if we return a failure state instead of panicking, we can never send using the same TxState struct (as the buffer is now permanently lost). There are 2 more analogous cases I can identify in the code. Should the code "fail fast" and panic (by calling expect) in these cases, or is it preferable to keep running in an inconsistent state?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

phil-levis commented:
I think reset should be OK above, but start performs an SPI operation (reads a register begin the start sequence) so needs DMA and the interrupt queue to have been initialized.

Phil

> On Sep 1, 2017, at 12:22 PM, ptcrews <notifications@github.com> wrote:
> 
> @ptcrews commented on this pull request.
> 
> In boards/imix/src/main.rs:
> 
> >      rf233.reset();
> -    rf233.set_pan(0xABCD);
> -    rf233.set_address(0x1008);
> -    //    rf233.config_commit();
> -
>      rf233.start();
> 
> So we're not entirely sure why, but the reset() and start() methods need to be below the creation of the chip otherwise the radio will not work, so I don't think they can be cleanly added to the current Mac initialization code. We could modify the Mac layer to expose a wrapper for reset() and start(), or we can try and figure out how to incorporate it more cleanly with the lower layers.
> 
> In either case, I'm not sure if we should implement it in this pull request, or submit another (small) pull request that modifies the Mac layer.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

alevy commented:
OK, I think it's possible to get rid of the remaining `expect`s, or at least restructure how they are called to make it a bit more readable. Are we passed the point where @ptcrews has cycles to fix things? If so I can maybe find time this month to patch it up, assuming I understand the flow correctly (as always more documentation of the intended functionality and internal fields would be useful...)

Otherwise I think this looks good to me.

ptcrews commented:
So I can work intermittently on this over the next two weeks, and I can attempt to remove the remaining `expect` calls or clean up the flow. I can certainly add more documentation and comments to make the code more readable. I'll aim for pushing the updated version by the end of the week.

alevy commented:
OK, @ptcrews, thanks. Let's see what you can wrap up by the end of the week and I'll take over from there. I think relying on you to do stuff in your spare time is a bad idea for everyone :). The PR is already in a good state, so worst case we'll just merge with perhaps a few small tweak from me.

ptcrews commented:
Sounds good to me!

ptcrews commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -0,0 +1,900 @@
+            // represents a significant bug if the packet is not there when
+            // transmission completes.
+            let mut packet =
+                self.packet.take().expect("Error: `packet` is None in call to end_transmit.");
```
First, the `packet` buffer for the `TxState` struct is statically allocated by the upper layers, and is passed to the `TxState` struct for the duration of the transmission. On the transmit path, ownership of this buffer is passed to `FragState::transmit_packet`, which then passes it to `TxState::init_transmit` (which gives it to the corresponding `TxState` struct). Thus, the `expect` call in `TxState::end_transmit` (line 407) makes sense; if we have lost this buffer, then we are in an inconsistent state, and the upper layer will never receive the buffer back. Since this must be a logic error, we should probably panic.

ptcrews commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -0,0 +1,900 @@
+            // occur - all other calls to `packet.take()` replace the packet,
+            // and thus the packet should always be here.
+            let mut buffer =
+                self.packet.take().expect("Error: `packet` is None in call to end_receive.");
```
Next, the `packet` buffer for the `RxState` struct is owned by the `RxState` struct, passed in via the `RxState::new` function. The `expect` call on line 536 thus makes sense, as if the `packet` buffer is no longer there, then the `RxState` struct would no longer be able to receive frames, and indicates a logic bug.

ptcrews commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -0,0 +1,900 @@
+                // Note that `tx_buf` should *always* be here, as we called
+                // `self.tx_buf.replace(..)` at the top of this function.
+                let tx_buf =
+                    self.tx_buf.take().expect("Error: `tx_buf` is None in send_done callback.");
```
The `expect` call on line 582 is for clarity; we replace the `tx_buf` buffer on line 567, so there is no way for this call to fail. Although the `expect` call could be removed by only calling `self.tx_buf.replace(..)` on control paths where `head.prepare_transmit_next_fragment` doesn't succeed, I believe this is the cleanest way to write this code.

ptcrews commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -0,0 +1,900 @@
+            // panic.
+            let mut frag_buf = self.tx_buf
+                .take()
+                .expect("Error: `tx_buf` is None in call to start_packet_transmit.");
```
The expect call inside `start_packet_transmit` (line 734) is again indicative of a logic bug. Since the `tx_buf` is owned by the `FragState` struct and passed between the radio and the fragmentation layer, if we are not busy and the `tx_buf` buffer is `None`, then we have lost the buffer, and can therefore never transmit.

ptcrews commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -0,0 +1,900 @@
+                // Otherwise, we are in an inconsistent state and can fail.
+                let mut packet = state.packet
+                    .take()
+                    .expect("Error: `packet` in RxState struct is `None` \
```
The final `expect` call is in the `receive_single_packet` function (line 810). Again, this buffer is owned by the `RxState` struct and if it is not there, we cannot receive any more frames. I had converted this buffer from a `TakeCell` to a `MapCell`, but this (again) touches on the "fail fast" debate, and whether to panic or return an error.

ptcrews commented:
So having looked at the code some more, I'm not sure how many `expect` calls can/should be removed. Although some can be avoided by using `map`, if the buffers are `None`, we are in an irrecoverable and inconsistent state. Below are the 5 `expect` calls in the code, with a quick explanation as to why I thought they were necessary. Note that these comments are also duplicated on their respective lines in the "Files changed" section on this pr.

1. First, the `packet` buffer for the `TxState` struct is statically allocated by the upper layers, and is passed to the `TxState` struct for the duration of the transmission. On the transmit path, ownership of this buffer is passed to `FragState::transmit_packet`, which then passes it to `TxState::init_transmit` (which gives it to the corresponding `TxState` struct). Thus, the `expect` call in `TxState::end_transmit` (line 407) makes sense; if we have lost this buffer, then we are in an inconsistent state, and the upper layer will never receive the buffer back. Since this must be a logic error, we should probably panic.

2. Next, the `packet` buffer for the `RxState` struct is owned by the `RxState` struct, passed in via the `RxState::new` function. The `expect` call on line 536 thus makes sense, as if the `packet` buffer is no longer there, then the `RxState` struct would no longer be able to receive frames, and indicates a logic bug.

3. The `expect` call on line 582 is for clarity; we replace the `tx_buf` buffer on line 567, so there is no way for this call to fail. Although the `expect` call could be removed by only calling `self.tx_buf.replace(..)` on control paths where `head.prepare_transmit_next_fragment` isn't called, I believe this is the cleanest way to write this code.

4. The expect call inside `start_packet_transmit` (line 734) is again indicative of a logic bug. Since the `tx_buf` is owned by the `FragState` struct and passed between the radio and the fragmentation layer, if we are not busy and the `tx_buf` buffer is `None`, then we have lost the buffer, and can therefore never transmit.

5. The final `expect` call is in the `receive_single_packet` function (line 810). Again, this buffer is owned by the `RxState` struct and if it is not there, we cannot receive any more frames. I had converted this buffer from a `TakeCell` to a `MapCell`, but this (again) touches on the "fail fast" debate, and whether to panic or return an error.

Overall, I can remove most of the `expect` calls by converting them to `map`s and using `unwrap_or`s, but since a failure at any of these points indicates an irrecoverable error, I'm not sure what the correct behavior should be. 

I also improved documentation and outlined the high-level control flow that should hopefully help, but I wanted to get an answer on whether to remove the `expect` calls and make them `map`s first or not.

alevy commented:
I rebased from master and fixed conflicts from the previous 6lowpan PR. It's worth retesting since I had to muck with the configuration a bit in `imix/src/main.rs` and while I think I got it right, I'm not 100% certain.

alevy commented:
I went through the `expect` calls and cleaned up the use of them a bit. However, high leveI, I think most of them can be obviated with better use of types. In particular, there are invariants like "packet should in RxState because busy is not true". That's a case for encoding this relationship in the type instead of simulating it with fields.

For example, `RxState` might have something like:

```
struct RxState {
    // no packet field or busy fields
   ...
   // instead
   idle: TakeCell<'static, [u8]>,
   ...
}
```

Now when we want to check for business we either map (or `take` + `match`, or `take` + `map`) over the `idle` field, in which case we'll just _have_ the packet if we're non-busy, or in other cases, where we just want the boolean we can call `is_none` (meaning busy) or `is_some` (meaning non-busy).

ptcrews commented:
Alright, thanks. I'll look over `imix/src/main.rs` and see if the tests still work. I'll also have a look at improving the use of types to avoid additional `expect` calls.

bbbert commented:
@ptcrews Before this gets merged, there are two things I think need to be fixed:
- Somehow `kernel/src/common/stream.rs` (along with additions to `kernel/src/common/mod.rs`) got re-introduced. This should be removed; the correct library to reference is now `capsules/src/net/stream.rs`.
- Somehow `capsules/src/radio.rs` also got re-introduced. Everything has since been moved to `capsules/src/ieee802154/*`.

I don't think any part of your code really touches these files, so you should be able to just remove it.

ptcrews commented:
@alevy If it's clean enough, I think that just calling `static_init!` in the test suite itself would probably be the best idea for usability.

@bbbert I'm not sure how that happened, but I'll remove them

ptcrews commented:
@alevy I managed to get the test suite up and running, and I only found a couple minor issues with the current code. I'll push a fix + better test suite initialization later today. I'm still working on using types to avoid the `expect` calls, so that won't get done until later.

ptcrews commented:
I moved the initialization code to the `initialize_all` function inside `lowpan_frag_test.rs`. It didn't make the initialization any cleaner per se, but it did remove the complexity from the `main.rs` file. Overall, the initialization routine could probably be cleaned up substantially, though I'm not sure exactly how at this point. 

I also removed the files mentioned by @bbbert - note that this required an additional modification to `capsules/src/lib.rs`, but everything still compiles.

Remaining tasks: Find a way to use types to avoid `expect` calls, finish final tests, clean-up and run rustfmt.

alevy commented on capsules/src/net/lowpan_fragment.rs:
```diff
@@ -570,14 +570,17 @@ impl<'a, A: time::Alarm> TxClient for FragState<'a, A> {
+                    // Note that we *must* replace the buffer before calling
+                    // end_packet_transmit, as it assumes we have the tx_buf
                     self.tx_buf.replace(tx_buf);
+                    self.end_packet_transmit(acked, retcode);
```
So this seems like the kind of situation where it makes sense to pass the buffer into `end_packet_transmit` instead, doesn't it?

alevy replied:
(but fine if we leave this sort of refactoring for later at this point)

ptcrews replied:
Yeah, it does. Not sure why I decided to separate them initially, but I just changed it.

alevy commented:
Confirm that I'm able to run the test and it works!
